### ISSUE TITLE
[BUG] Missing: Unity C# method for actually Minting a NFT #hackathon

### ISSUE THREAD

**Issue by yosun** [onflow/UnityFlowSDK#4]
### Current Behavior

The existing tutorials and documentation indicate that you are supposed to be able to mint a NFT using Flow. 

A `emulator_test_data.sh` script is provided to mint some test NFT, but this is using CLI and cadence. 

There needs to be a method for a Unity developer to call from Unity to mint a NFT. 

### Expected Behavior

A very simple snippet should be available. I assume this is asynchronous, but we can simply have a callback.

Please have a method that looks something like this available: 

````
MintNFT(string url, Callback); 
````

And the `Callback(string result, int resource)` returns the resource and success/error when minted (replace with appropriate arguments - i have only a high level understanding of how to mint, but as I understand it is a integer resource that is returned

Where `url` is supposed by the Unity dev, that links to the thing to be minted.

### Steps To Reproduce

There needs to be a method for a Unity developer to call from Unity to mint a NFT. 

### Environment

```markdown
All
```

### What are you currently working on that this is blocking?

an awesome AR x AI mobile app for the hackathon

**Comment by Executus** [comment]
A `MintNFT` function is not something that we would put in the Unity SDK for multiple reasons. 

Firstly, it is too application specific. There could be many games integrated with the Flow blockchain that do not use NFTs, and those that do are going to define their own NFTs. It would not be possible for the SDK to be "aware" of a game's NFTs to mint them. To give an example in the non-blockchain space, it would be like Unity having a `MintGold` function. There are many games made in Unity that do not have Gold, so this function would be useless to them. Unity could also provide `MintSilver`, `MintCopper`, `MintDollars` and `MintFood` functions, which might be useful to some games, but not the majority. These are game-specific elements that the game should implement, not Unity. It's the same for the Flow SDK and `MintNFTs`. If your game has NFTs called Collectibles, it would make sense for your game to have a `MintCollectibles` function, but it doesn't make sense for the SDK to have this. 

Secondly, NFTs must be minted from cadence, in a transaction signed by a Flow account. It cannot be done directly from Unity. We could write a helper function that submits a transaction to do this, but this would lead to complexity around who signs the transaction, what happens if the transaction fails etc, and the solutions to these problems would differ from game to game. This leads to the first point - we'd be making game-specific functionality in the SDK, which is not what the SDK should be doing. 

What is actually needed are clear examples and tutorials of how to mint NFTs, in the context of a game. The Flow Words sample  has been modified in the `pre-releases/v2.0.0` branch to use testnet, so it no longer uses the `emulator_test_data` script you mentioned. We will continue to create new samples and tutorials to provide more examples of how to use the Flow SDK, including how to mint NFTs.

**Comment by yosun** [comment]
I may be oversimplifying, but (correct me if I am misunderstanding) it seems, a classic NFT is just a resource that is tied to a url string (often pointing to a image)

So, you could just have a very general catch-all `MintNFT(int id,string url)` method that does not care if the URL represents gold or food or whatever - it is simply a URL (string) tied to a flow resource id (int). The app then processes it (the wallet usually just loads the image url). 

Or, `int MintNFT(string url)` - and it returns the resource id. 
But, as I understand it is asynchronous,  so this perhaps makes more sense: `MintNFT(string url,CallbackLambda)` where once the item is minted, the id is passed to the `CallbackLambda`. 

So effectively, the game logic stays within the game (whether to mint gold or whatever, and to pass in the resulting URL to   `MintNFT`). `MintNFT()` - a very common generalized operation that many other apps/games could use - is provided via SDK.

**Comment by Executus** [comment]
`it seems, a classic NFT is just a resource that is tied to a url string`
An NFT on Flow is even more simple than that - it is just a unique id (UInt64). That is the only required field. You can define any other fields you like (such as a string for url) in the definition of your NFT, but `id` is the only required field for all NFTs. Here is the cadence interface that all NFTs should implement: 

https://github.com/onflow/flow-nft/blob/master/contracts/NonFungibleToken.cdc

Regardless of what we put in the SDK, the developer will need to define their NFT in their own contract and deploy the contract to a Flow account. If you have not done so already, I highly recommend you take a look at the Flow Playground tutorials: 

https://play.flow.com

In particular this tutorial, which steps through how to define and mint NFTs in cadence: 

https://developers.flow.com/cadence/tutorial/05-non-fungible-tokens-1

In a future release of the Flow SDK for Unity, we will mirror the Flow Playground tutorials with Unity sample tutorials.

**Comment by yosun** [comment]
Yes, I did look through the Create NFT playground tutorial before posting here. 

I tried modifying it a bit but then had a lot of trouble just writing a `getNFTMetadata` method in cadence and then realized I was not sure how to reference the NFT we just created. That playground example's NFTExist script only checked if something was written in storage. This does not really help because in practice, we need to be able to lookup what was written to storage. 

```
pub contract BasicNFT {

    // Declare the NFT resource type
    pub resource NFT {
        // The unique ID that differentiates each NFT
        pub let id: UInt64

        // String mapping to hold metadata
        pub var metadata: {String: String}

        // Initialize both fields in the init function
        init(initID: UInt64,url: String) {
            self.id = initID
            self.metadata =  {"url": url}
        }
 
    }

    // Function to create a new NFT
    pub fun createNFT(id: UInt64,url: String): @NFT {
        return <-create NFT(initID: id,url: url)
    }
   /* pub fun getNFTMetadata(id: UInt64): @NFT {
        // ?? How do we reference a previously created NFT and get its metadata? 
    }*/

    // Create a single new NFT and save it to account storage
    init(  ) {
        self.account.save<@NFT>(<-create NFT(initID: 1,url: "https://napkinmatic.com"), to: /storage/BasicNFTPath)
    }
}
```

Well, actually, I just figured out how to load metadata in cadence - I am not sure if this is the best and only way, where I have to save it to avoid "loss of resource" error? 

Also, I am not sure how to load another NFT with a different ID. 

**Question: Do NFTs have to be stored to a different storage path? I had thought defining the resource ID would serve as a primary key (?)** 
```
import BasicNFT from 0x01

/// ?? Load metadata but somehow we have to save it to avoid loss of resource (?) 

transaction {
    prepare(signer1: AuthAccount) {
             
        // Load the NFT from signer1's account
        let nft <- signer1.load<@BasicNFT.NFT>(from: /storage/BasicNFTPath)
            ?? panic("Could not load NFT")

        log(nft.metadata);

        // Save the NFT to signer2's account
       signer1.save<@BasicNFT.NFT>(<-nft, to: /storage/BasicNFTPath)

    }
}
```

**Comment by yosun** [comment]
(Cadence often gives me a headache, the syntax is so foreign - so I apologize for these quick replies.) 

Effectively, what is needed from Unity, that would universally help Unity developers use Flow, is a very simple Cadence.Read/Write() version of PlayerPrefs.SetString/Int/etc 
https://docs.unity3d.com/ScriptReference/PlayerPrefs.html

For example:

```
void Cadence.WriteString(string stringName,string stringValue);
string Cadence.ReadString(string stringName); 
```

where `stringName` can also just be Uint64 resource ID or other unique identifiers

**Comment by Executus** [comment]
Could you please post the cadence question in the `cadence` channel on discord? There are much better experts on cadence than myself who can answer, plus it's very useful for the community to see the answers to your question. 

https://discord.gg/flow

Regarding your latter post, the best way to think about this is in terms of cadence Scripts and Transactions. 

Reading data = execute a Script
Writing data = submit a Transaction (must be signed by an account)

Reading any data from the blockchain is nice and simple (once you get to know cadence syntax). You write a Script (must have a `main` function), which can take arguments and can optionally return whatever data you want. It could be something simple like a string or an int, or something as complex as an array of nested structs. Use the SDK's `Scripts.ExecuteAtLatestBlock` to do this. 

Writing to the blockchain is much more complicated - the cadence code must be submitted in a Transaction, which is signed by the account/s that's going to be affected. From a Web3 point of view - you're trying to modify someone's data, so you need their permission. Like Scripts, Transactions take some cadence code (though the format is different - they don't have a `main` function), and can optionally take some arguments. Unlike Scripts though, they do not return anything, and they take a lot longer to execute. Use the SDK's `Transactions.Submit` to do this. 

In the pre-release of v2.0.0, we have introduced a new feature called Cadence Convert. This allows you to convert your script\transaction arguments, and script return data, between C# and cadence very easily. The document for this feature is in `Documentation~/convert`. Have a read of this, as it may help you with what you're trying to do.

**Comment by Niosop** [comment]
> Question: Do NFTs have to be stored to a different storage path? I had thought defining the resource ID would serve as a primary key (?)

Take a look at the Collection portion of the NonFungibleToken contract.  A user will typically create a collection resource (the dev would implement something like the createEmptyCollection function which creates and returns a new collection resource) and store it at a unique known location in private storage.  Often this unique location is defined in the contract.  Then it would link the public capability CollectionPublic of that collection to a unique known public path.  This will allow anyone to list the NFTs of that type held by the collection, deposit an NFT of that type into the collection, or borrow a reference to an NFT to read its data without actually taking possession of that NFT (using the id to tell it which NFT in the collection it wants a reference to).  On Flow, the NFTs are held in the owner's account, not in the contract.
